File: solutions/123/main.py
from typing import List 

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # dp table 
        # 0 = no action 
        # 1 = 1st buy 
        # 2 = 1st sell 
        # 3 = 2nd buy 
        # 4 = 2nd sell 
        dp = [[0] * 5 for _ in range(len(prices))]
        # initialize 
        dp[0][0] = 0 
        dp[0][1] = -prices[0]
        dp[0][2] = 0 
        dp[0][3] = -prices[0]
        dp[0][4] = 0 
        # traverse 
        for i in range(1, len(prices)): 
            dp[i][0] = dp[i-1][0]
            dp[i][1] = max(dp[i-1][1], dp[i-1][0] - prices[i])
            dp[i][2] = max(dp[i-1][2], dp[i-1][1] + prices[i])
            dp[i][3] = max(dp[i-1][3], dp[i-1][2] - prices[i])
            dp[i][4] = max(dp[i-1][4], dp[i-1][3] + prices[i])
        return max(dp[-1][2], dp[-1][-1])
        
if __name__ == "__main__": 
    prices = [7,6,4,3,1]
    s = Solution()
    assert s.maxProfit(prices) == 0

File: solutions/123/main.rs
struct Solution; 

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let n = prices.len();
        let mut dp = vec![vec![0; 5]; n]; 

        // no action 
        dp[0][0] = 0;
        // 1st buy 
        dp[0][1] = -prices[0]; 
        // 1st sell 
        dp[0][2] = 0; 
        // 2nd buy 
        dp[0][3] = -prices[0]; 
        // 2nd sell 
        dp[0][4] = 0; 

        for i in 1..n {
            dp[i][0] = dp[i-1][0]; 
            // either already buy at i-1, or buy at i 
            dp[i][1] = std::cmp::max(dp[i-1][1], dp[i-1][0]-prices[i]);
            dp[i][2] = std::cmp::max(dp[i-1][2], dp[i-1][1]+prices[i]);
            dp[i][3] = std::cmp::max(dp[i-1][3], dp[i-1][2]-prices[i]); 
            dp[i][4] = std::cmp::max(dp[i-1][4], dp[i-1][3]+prices[i]); 
        }

        std::cmp::max(dp[n-1][1], dp[n-1][4])
    }
}